### PR TITLE
Adding Century Link Cloud Provider

### PR DESCRIPTION
**Pull request by ablinkinz**
created a new cloud provider to support Century Link Cloud. More features coming to better match the GUI operations they provide

### What does this PR do?
Adds the Century Link Cloud Provider to allow salt-cloud provisioning to CLC using their python sdk

### Tests written?

No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

File: salt/cloud/clouds/clc.py [hunk 1]
```diff
@@ -0,0 +1,368 @@
+# -*- coding: utf-8 -*-
+'''
+CenturyLink Cloud Module
+===================
+
+.. versionadded:: 2017.7.14
```
Comment (cachedout) [inline salt/cloud/clouds/clc.py]: This will be `Oyxgen` instead

File: salt/cloud/clouds/clc.py [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+
+    pip install clc-sdk
+
+.. note::
+  For sdk reference see: https://github.com/CenturyLinkCloud/clc-python-sdk
+
+Configuration
+=============
+
+To use this module, set up the clc-sdk, username,password,key in the
```
Comment (cachedout) [inline salt/cloud/clouds/clc.py]: spaces between commas? also, this references `user` and this says `username`. Should they be consistent?

File: salt/cloud/clouds/clc.py [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+cloud configuration at
+``/etc/salt/cloud.providers`` or ``/etc/salt/cloud.providers.d/clc.conf``:
+
+.. code-block:: yaml
+
+    my-clc-config:
+      driver: clc
+      user: 'web-user'
```
Comment (cachedout) [inline salt/cloud/clouds/clc.py]: Are quotes necessary here?

File: salt/cloud/clouds/clc.py [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+      password: 'verybadpass'
```
Comment (cachedout) [inline salt/cloud/clouds/clc.py]: And here?

File: salt/cloud/clouds/clc.py [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+import logging
+import time
+import os.path
+import subprocess
+import json
+# Import salt libs
+import salt.utils
+import salt.utils.cloud
+import salt.utils.xmlutil
+#import pip
```
Comment (cachedout) [inline salt/cloud/clouds/clc.py]: Please remove this line if it isn't being used.

File: salt/cloud/clouds/clc.py [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+import importlib
+from salt.exceptions import SaltCloudSystemExit
+from flask import Flask, request
```
Comment (cachedout) [inline salt/cloud/clouds/clc.py]: `flask` is not a dep of salt. This will raise an `ImportError` in most cases.

File: salt/cloud/clouds/clc.py [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+# Import salt cloud libs
+import salt.config as config
+import requests
+# Attempt to import clc-sdk lib
+try:
+        importlib.import_module('clc')
```
Comment (cachedout) [inline salt/cloud/clouds/clc.py]: I'm not clear on why this line is necessary. Could you please clarify?

File: salt/cloud/clouds/clc.py [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+    import salt.ext.six as six
+    HAS_SIX = True
+except ImportError:
+    # Salt version <= 2014.7.0
+    try:
+        import six
+    except ImportError:
+        HAS_SIX = False
+
+IP_RE = r'^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$'
```
Comment (cachedout) [inline salt/cloud/clouds/clc.py]: I don't see this being used anywhere?

File: salt/cloud/clouds/clc.py [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+
+# Only load in this module if the CLC configurations are in place
+def __virtual__():
+    '''
+    Check for CLC configuration and if required libs are available.
+    '''
+    if get_configured_provider() is False:
+       return False
+
+    if get_dependencies() is False:
```
Comment (cachedout) [inline salt/cloud/clouds/clc.py]: This could be combined with the above conditional with an `or`

File: salt/cloud/clouds/clc.py [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+    password = config.get_cloud_config_value(
+    'password', get_configured_provider(), __opts__, search_global=False
+    )
+    token = config.get_cloud_config_value(
+    'token', get_configured_provider(), __opts__, search_global=False
+    )
+    token_pass = config.get_cloud_config_value(
+    'token_pass', get_configured_provider(), __opts__, search_global=False
+    )
+    creds = {"user":user,"password":password,"token":token,"token_pass":token_pass}
```
Comment (cachedout) [inline salt/cloud/clouds/clc.py]: Please add spaces here after the commas and use single-quotes, per our style guide.

File: salt/cloud/clouds/clc.py [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+    )
+
+    # Bootstrap!
+    ret = __utils__['cloud.bootstrap'](vm_, __opts__)
+    return_message = {"Server Name":name,"IP Address":vm_['ssh_host']}
+    ret.update(return_message)
+    return(return_message)
+
+def destroy(name, call=None):
+    print("destroying ")
```
Comment (cachedout) [inline salt/cloud/clouds/clc.py]: Please don't print from here. We need to return True instead, most likely. We could also log if desired.

**Review by cachedout — CHANGES_REQUESTED** [review]

File: salt/cloud/clouds/clc.py (hunk 7 above)
Comment (cachedout) [reply salt/cloud/clouds/clc.py]: Thanks for the clarification. This is super-weird, though. Is an upstream bug filed with the `clc` maintainers?

Comment (ablinkinz) [reply salt/cloud/clouds/clc.py]: I submitted a ticket to them via our enterprise account....no updates on it yet.

File: salt/cloud/clouds/clc.py (hunk 3 above)
Comment (ablinkinz) [reply salt/cloud/clouds/clc.py]: usernames/tokens can have special characters that can cause issues with python. I ran into a couple instances in our environment

File: salt/cloud/clouds/clc.py (hunk 4 above)
Comment (ablinkinz) [reply salt/cloud/clouds/clc.py]: usernames/tokens can have special characters that can cause issues with python. I ran into a couple instances in our environment

File: salt/cloud/clouds/clc.py (hunk 3 above)
Comment (cachedout) [reply salt/cloud/clouds/clc.py]: Ah, K. Thanks for the clarification.

**Review by cachedout — CHANGES_REQUESTED** [review]
The linter is requesting a number of changes so that this conforms to our coding style. The list of requested changes is here:

https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/15317/violations/file/salt/cloud/clouds/clc.py/

**Comment by ablinkinz** [comment]
@cachedout I think I got it all fixed up now. I lined up my indentions, added the whitespaces after commas and such....there were a few that showed white space requirements after a comma that actually did have the whitespace. The same is true for some trailing spaces.
Thanks for your help sir

**Comment by ablinkinz** [comment]
@cachedout I finally got it all cleaned up and running with a beautiful green check mark next to the linter :) should ACTUALLY be good now. 

Thanks again for your patience and help!

**Comment by cachedout** [comment]
@ablinkinz Looks great now. Let's get this in!

**Review by cachedout — APPROVED** [review]

**Pull request merged**